ccm_web/server/src/lti/lti.service.ts

```diff
@@ -113,6 +113,10 @@ export class LTIService implements BeforeApplicationShutdown {
       })
     })
 
+    provider.onInvalidToken(async (req: Request, res: Response) => {
+      return res.redirect('/launch-error')
```

Comment:
Can you tell the difference between launch from safari vs student launching CCM (by somehow grabbing url)? doesn't it both invalid token launches. 

What you have right now will work and not asking to detect safari, just want to know if there is a way LTIJS is giving that info.

Reply:
I think if it's a student, the launch still works, we just don't log them in and redirect them to `access-denied`. See `onConnect` hook.

Reply:
I looked it up a bit more. I see that `LTI login/authorize/Launch` is happening if 3rd-party cookies are disabled. some how the code block `onInvalidToken` is detecting as cookies are disabled and tool launch is failed. So I did not see from the code either (documentation of LTIJS is not much about this) much description about this function. 

I would just go with an understanding that **onInvalidToken** is written for seeing cookies disable or enabled. IF that is not true then we can find out based on the error case that comes to us in future.